File: lib_common/src/main/java/com/fly/tour/common/view/CommonDialogFragment.java
package com.fly.tour.common.view;

import android.content.DialogInterface;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentManager;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.fly.tour.common.R;
import com.fly.tour.common.util.DisplayUtil;
import com.fly.tour.common.util.log.KLog;

public class CommonDialogFragment extends DialogFragment {
    public static final String TAG = CommonDialogFragment.class.getSimpleName();
    private OnDialogClickListener mOnDialogClickListener;
    private static boolean isShowing = false;//避免弹多个dialog
    @Override
    public void dismiss() {
        super.dismiss();
        isShowing = false;
        KLog.v(TAG,"dismiss start...");
    }

    @Override
    public void show(FragmentManager manager, String tag) {
        super.show(manager, tag);
        isShowing = true;
    }

    public static boolean isShowing() {
        return isShowing;
    }
    public CommonDialogFragment setOnDialogClickListener(OnDialogClickListener onDialogClickListener) {
        mOnDialogClickListener = onDialogClickListener;
        return this;
    }

    public interface OnDialogClickListener {
        void onLeftBtnClick(View view);
        void onRightBtnClick(View view);
    };
    public static CommonDialogFragment newInstance(Builder builder){
        CommonDialogFragment commonDialogFragment = new CommonDialogFragment();
        Bundle args = new Bundle();
        args.putString("title",builder.title);
        args.putString("describe",builder.describe);
        args.putString("leftbtn",builder.leftbtn);
        args.putString("rightbtn",builder.rightbtn);
        args.putInt("rightbtncolor", builder.btnRightTextColor);
        commonDialogFragment.mOnDialogClickListener = builder.mListener;
        commonDialogFragment.setArguments(args);
        return commonDialogFragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onStart() {
        super.onStart();
        getDialog().getWindow().setLayout(getResources().getDisplayMetrics().widthPixels - DisplayUtil.dip2px(40) * 2, ViewGroup.LayoutParams.WRAP_CONTENT);
        getDialog().getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_common_dialog, container, false);
        Bundle arguments = getArguments();
        String title = null;
        String describe = null;
        String leftbtn = null;
        String rightbtn = null;
        int rightBtnTextColor = 0;
        if (arguments != null) {
            title = arguments.getString("title");
            describe = arguments.getString("describe");
            leftbtn = arguments.getString("leftbtn");
            rightbtn = arguments.getString("rightbtn");
            rightBtnTextColor = arguments.getInt("rightbtncolor", 0);
        }
        TextView txtTitle = view.findViewById(R.id.txt_common_dialog_title);
        TextView txtDescribe = view.findViewById(R.id.txt_common_dialog_describe);
        Button btnLeft = view.findViewById(R.id.btn_common_dialog_left);
        Button btnRight = view.findViewById(R.id.btn_common_dialog_right);
        View btnHalving = view.findViewById(R.id.view_halving_line);
        if (!TextUtils.isEmpty(title)) {
            txtTitle.setText(title);
        }else{
            txtTitle.setVisibility(View.GONE);
        }
        if(!TextUtils.isEmpty(describe)){
            txtDescribe.setText(describe);
        }
        if (!TextUtils.isEmpty(leftbtn)) {
            btnHalving.setVisibility(View.VISIBLE);
            btnLeft.setVisibility(View.VISIBLE);
            btnLeft.setText(leftbtn);
        } else {
            btnHalving.setVisibility(View.GONE);
            btnLeft.setVisibility(View.GONE);
        }
        if (!TextUtils.isEmpty(rightbtn)) {
            btnRight.setText(rightbtn);
        }
        if (rightBtnTextColor != 0) {
            btnRight.setTextColor(rightBtnTextColor);
        }
        btnLeft.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(mOnDialogClickListener != null){
                    mOnDialogClickListener.onLeftBtnClick(view);
                }
                dismiss();
            }
        });
        btnRight.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                if(mOnDialogClickListener != null){
                    mOnDialogClickListener.onRightBtnClick(view);
                }
                dismiss();
            }
        });
        return view;
    }
    public static class  Builder{
        String title;
        String describe;
        String leftbtn;
        String rightbtn;
        int btnRightTextColor;
        OnDialogClickListener mListener;

        public Builder() {
        }


        public Builder setTitle(String title) {
            this.title = title;
            return this;
        }
        public Builder setDescribe(String describe) {
            this.describe = describe;
            return this;
        }
        public Builder setLeftbtn(String leftbtn) {
            this.leftbtn = leftbtn;
            return this;
        }
        public Builder setRightbtn(String rightbtn) {
            this.rightbtn = rightbtn;
            return this;
        }
        public Builder setOnDialogClickListener(OnDialogClickListener listener) {
            this.mListener = listener;
            return this;
        }

        public Builder setRightbtnTextColor(int rightBtnTextcolor) {
            this.btnRightTextColor = rightBtnTextcolor;
            return this;
        }

        public CommonDialogFragment build() {
            return CommonDialogFragment.newInstance(this);
        }
    }

    @Override
    public void onCancel(DialogInterface dialog) {
        super.onCancel(dialog);
        isShowing = false;
        KLog.v(TAG,"onCancel start...");
    }
}

File: lib_common/src/main/java/com/fly/tour/common/view/CommonDialogFragment.kt
package com.fly.tour.common.view

import android.content.DialogInterface
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView

import com.fly.tour.common.R
import com.fly.tour.common.util.DisplayUtil
import com.fly.tour.common.util.log.KLog

class CommonDialogFragment : androidx.fragment.app.DialogFragment() {
    private var mOnDialogClickListener: OnDialogClickListener? = null
    override fun dismiss() {
        super.dismiss()
        isShowing = false
        KLog.v(TAG, "dismiss start...")
    }
    override fun show(manager: FragmentManager, tag: String?) {
        if (manager != null) {
            super.show(manager, tag)
        }
        isShowing = true
    }

    fun setOnDialogClickListener(onDialogClickListener: OnDialogClickListener): CommonDialogFragment {
        mOnDialogClickListener = onDialogClickListener
        return this
    }

    interface OnDialogClickListener {
        fun onLeftBtnClick(view: View)
        fun onRightBtnClick(view: View)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onStart() {
        super.onStart()
        dialog!!.window!!.setLayout(resources.displayMetrics.widthPixels - DisplayUtil.dip2px(40f) * 2, ViewGroup.LayoutParams.WRAP_CONTENT)
        dialog!!.window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_common_dialog, container, false)
        val arguments = arguments
        var title: String? = null
        var describe: String? = null
        var leftbtn: String? = null
        var rightbtn: String? = null
        var rightBtnTextColor = 0
        if (arguments != null) {
            title = arguments.getString("title")
            describe = arguments.getString("describe")
            leftbtn = arguments.getString("leftbtn")
            rightbtn = arguments.getString("rightbtn")
            rightBtnTextColor = arguments.getInt("rightbtncolor", 0)
        }
        val txtTitle = view.findViewById<TextView>(R.id.txt_common_dialog_title)
        val txtDescribe = view.findViewById<TextView>(R.id.txt_common_dialog_describe)
        val btnLeft = view.findViewById<Button>(R.id.btn_common_dialog_left)
        val btnRight = view.findViewById<Button>(R.id.btn_common_dialog_right)
        val btnHalving = view.findViewById<View>(R.id.view_halving_line)
        if (!TextUtils.isEmpty(title)) {
            txtTitle.text = title
        } else {
            txtTitle.visibility = View.GONE
        }
        if (!TextUtils.isEmpty(describe)) {
            txtDescribe.text = describe
        }
        if (!TextUtils.isEmpty(leftbtn)) {
            btnHalving.visibility = View.VISIBLE
            btnLeft.visibility = View.VISIBLE
            btnLeft.text = leftbtn
        } else {
            btnHalving.visibility = View.GONE
            btnLeft.visibility = View.GONE
        }
        if (!TextUtils.isEmpty(rightbtn)) {
            btnRight.text = rightbtn
        }
        if (rightBtnTextColor != 0) {
            btnRight.setTextColor(rightBtnTextColor)
        }
        btnLeft.setOnClickListener { view ->
            if (mOnDialogClickListener != null) {
                mOnDialogClickListener!!.onLeftBtnClick(view)
            }
            dismiss()
        }
        btnRight.setOnClickListener { view ->
            if (mOnDialogClickListener != null) {
                mOnDialogClickListener!!.onRightBtnClick(view)
            }
            dismiss()
        }
        return view
    }

    class Builder {
        internal lateinit var title: String
        internal lateinit var describe: String
        internal lateinit var leftbtn: String
        internal lateinit var rightbtn: String
        internal var btnRightTextColor: Int = 0
        internal lateinit var mListener: OnDialogClickListener


        fun setTitle(title: String): Builder {
            this.title = title
            return this
        }

        fun setDescribe(describe: String): Builder {
            this.describe = describe
            return this
        }

        fun setLeftbtn(leftbtn: String): Builder {
            this.leftbtn = leftbtn
            return this
        }

        fun setRightbtn(rightbtn: String): Builder {
            this.rightbtn = rightbtn
            return this
        }

        fun setOnDialogClickListener(listener: OnDialogClickListener): Builder {
            this.mListener = listener
            return this
        }

        fun setRightbtnTextColor(rightBtnTextcolor: Int): Builder {
            this.btnRightTextColor = rightBtnTextcolor
            return this
        }

        fun build(): CommonDialogFragment {
            return CommonDialogFragment.newInstance(this)
        }
    }

   override fun onCancel(dialog: DialogInterface) {
        if (dialog != null) {
            super.onCancel(dialog)
        }
        isShowing = false
        KLog.v(TAG, "onCancel start...")
    }

    companion object {
        val TAG = CommonDialogFragment::class.java!!.getSimpleName()
        var isShowing = false
            private set//避免弹多个dialog

        fun newInstance(builder: Builder): CommonDialogFragment {
            val commonDialogFragment = CommonDialogFragment()
            val args = Bundle()
            args.putString("title", builder.title)
            args.putString("describe", builder.describe)
            args.putString("leftbtn", builder.leftbtn)
            args.putString("rightbtn", builder.rightbtn)
            args.putInt("rightbtncolor", builder.btnRightTextColor)
            commonDialogFragment.mOnDialogClickListener = builder.mListener
            commonDialogFragment.arguments = args
            return commonDialogFragment
        }
    }
}
